fix(discover): Don't include cursor headers when unused

wmak opened the pull request:
- This is just a small optimization to reduce the header response
  because of an issue with large headers, it's goal is just to reduce
  the headers a bit, but doesn't address the main issue yet.
  - The actual solution is going to require us to add reasonable limits
    on the query length and possibly limit our number of fields more
    along with introducing special endpoints when we require more fields
    than the limit (of 20 today)

wmak committed:
fix(discover): Don't include cursor headers when unused

- This is just a small optimization to reduce the header response
  because of an issue with large headers, it's goal is just to reduce
  the headers a bit, but doesn't address the main issue yet.
  - The actual solution is going to require us to add reasonable limits
    on the query length and possibly limit our number of fields more
    along with introducing special endpoints when we require more fields
    than the limit (of 20 today)

markstory commented on src/sentry/api/base.py:
```diff
@@ -327,7 +327,11 @@ def paginate(
+
+        # Don't include cursor headers if the client won't be using them
+        if not request.GET.get("no_cursor_headers"):
+            self.add_cursor_headers(request, response, cursor_result)
```
Should this be hoisted to the endpoint that is using it? We could skip pagination entirely that way.

wmak replied:
I considered that, but thought putting it here would mean that if we saw other endpoints bumping up against the header limit they'd be able to use this parameter as well. Though thinking about it a bit more, I'm not sure how often we'll have:
a) An endpoints that's normally paginated
b) But that for a use case we you only want a single page for
c) And because of various reasons is at the header limit

so yeah 🤔 I'll hoist 🧗 this up, and if we notice this happening on other endpoints we can reconsider moving it down again.

markstory replied:
>  and if we notice this happening on other endpoints we can reconsider moving it down again.

Easier to refactor to an abstraction later over adding a leaky/incorrect abstraction to early.

wmak replied:
:+1: Another good reason to do it

wmak committed:
ref: Hoist no pagination up to the events endpoint

- Instead of adding this everywhere, only add it to the events endpoint
- Add a function to get the per_page parameter so that we can re-use it
  in the org events endpoint

markstory commented on src/sentry/static/sentry/app/utils/discover/genericDiscoverQuery.tsx:
```diff
@@ -153,6 +159,9 @@ class GenericDiscoverQuery<T, P> extends React.Component<Props<T, P>, State<T>>
       apiPayload.per_page = limit;
     }
+    if (noPagination) {
+      apiPayload.no_pagination = noPagination;
```
Most of the API parameters are camelBacked as well. So `noPagination` in this case.

Zylphrex commented:
@k-fish I think you'll want to enable this option for the query you're making to generate the colour bar on the performance landing page.

wmak committed:
ref: rename no_pagination to noPagination
